### ISSUE TITLE
Explore a non-parametrical version of the wf-psf model

### ISSUE THREAD

**Issue by tobias-liaudat** [CosmoStat/wf-psf#12]
Try out if we can add an additive non-parametric layer to the OPD. 

Later we could see if it is worth to add a non-parametric layer at the pixel level so that it can cope with effects happening at the pixel level. Nevertheless, this would mean that we should have a dataset that include these effects.

How can we represent the variability of the PSF?
- At first, we can go for a classic matrix factorisation in a RCA-style, a linear model. Learning a feature matrix (S) and a variability matrix (A) that contains the varying weights of each feature.

How to validate this approach?
- Not that simple given the fact that the PSF simulation tool is generating PSFs following polynomial variations of the Zernike coefficients. One way to go would be to simulate using a high number of Zernike polynomials (ex 45) and then set a limited number of Zernike polynomials in the parametric part of our wf-psf model (ex 10). Then evaluate the performance of the parametric model (only estimating Zernike coefficients) and the semi-parametric model (estimating the Zernike coefficients and the matrix factorisation).

**Comment by tobias-liaudat** [comment]
## To Do

- [x] First semi-parametric model using the factorisation S alpha Pi (where we use a polynomial variation (for simplicity right now)) [commit here](https://github.com/tobias-liaudat/wf-psf/commit/8b3787f4b71a6df6eac35c1e9c86ddada6469370)
- [x] Generate complex dataset, ie 45 Zernike order. [commit here](https://github.com/tobias-liaudat/wf-psf/commit/0f7c4d795212c292ab67f0fef906ae4c61f927e6)
- [x] Try to estimate it with the parametric PSF model with a lower order, ie 15.
- [x] Evaluate the errors made. [commit here](https://github.com/tobias-liaudat/wf-psf/commit/dfdbd32f30c69872a0184577c2e56cfc3d74164f)
- [x] See if the non-parametric layer is able to improve the results in this set-up.

For this I might need to define a _custom training_. First optimise the parametric part with the non-parametric part set to zero. At this point initialise the non-parametric model to a non-zero value. Then fix the parametric part and optimise over the non-parametric part.

To consider:
- Add a regularisation to the non parametric part, ie L1 norm to alpha and a denoising strategy to S columns (eigenOPDs).


## Next step

After this done, the next step is to define a non-parametric model that does not have polynomial variation. This would imply that we need to handle the batch indexes. Depending on which indexes the batch uses we are going to optimise over certain columns of the _A_ matrix (MCCD nomenclature).

**Comment by tobias-liaudat** [comment]
## First results

Trying to model a complex dataset, Zernike order 45, with a lower number of Zernike polynomials. Both trained in 20 epochs.

## Dataset z=45, model z=15

<img width="772" alt="model_15_psfs" src="https://user-images.githubusercontent.com/16106854/106290107-5e0dd980-624a-11eb-9009-57e7f16cdeb6.png">

<img width="1103" alt="model_15_mat" src="https://user-images.githubusercontent.com/16106854/106290116-6403ba80-624a-11eb-8f17-b23fea00629f.png">


## Dataset z=45, model z=45

<img width="768" alt="model_45_psfs" src="https://user-images.githubusercontent.com/16106854/106290250-8990c400-624a-11eb-9365-75c7ac1deb4d.png">

<img width="976" alt="model_45_mat" src="https://user-images.githubusercontent.com/16106854/106290262-8dbce180-624a-11eb-938d-a8e96ef01d00.png">

## Quantitative results

| Model z order      | train RMSE | test RMSE  |
| :------: | :--------: | :--------: |
| 15        |  2.7169e-04 | 2.5854e-04 |
| 45       | 2.0343e-04 | 2.0001e-04 |

**Comment by tobias-liaudat** [comment]
## First results using a naive non-parametric model

Using polynomial variations right now for the eigenOPDs.

Details of the model:
Parametric: Adam , lr=1e-2, epochs=20, **15** Zernike order, max polynomial variation degree is **2**
Non-parametric: Adam , lr=1.0, epochs=100, max polynomial variation degree is **3**

Details of the dataset: 
**45** Zernike order, max polynomial variation degree is **2**
 
## Quantitative 

| Model z order      | train RMSE | test RMSE  |
| :------: | :--------: | :--------: |
| 15        |  2.7169e-04 | 2.5854e-04 |
| 45       | 2.0343e-04 | 2.0001e-04 |
| 15 + nonparam        |  9.5326e-05 |  9.3800e-05 |

## Qualitative 

Learned eigenOPDS

<img width="1192" alt="Screenshot 2021-02-01 at 12 04 27" src="https://user-images.githubusercontent.com/16106854/106460685-ffd43700-6493-11eb-8633-ee7c093bcb7f.png">

Learned alpha matrix

<img width="302" alt="Screenshot 2021-02-01 at 13 43 40" src="https://user-images.githubusercontent.com/16106854/106460692-02cf2780-6494-11eb-8491-0edd4abedd69.png">

Loss function of parametric estimation

<img width="836" alt="Screenshot 2021-02-01 at 13 49 51" src="https://user-images.githubusercontent.com/16106854/106460973-6d806300-6494-11eb-9c59-7f2f4f9fe915.png">


Loss function of non-parametric estimation

<img width="823" alt="Screenshot 2021-02-01 at 13 49 26" src="https://user-images.githubusercontent.com/16106854/106460991-72451700-6494-11eb-8034-0a61a94e830e.png">

**Comment by tobias-liaudat** [comment]
Before closing this issue I'll share more plots to analyse the first results comparing:

- Ground truth model with Z=45
- SemiParametric model with Z=15
- Parametric model with Z=45

It's interesting to see that:
- even if the resulting OPD from the semi-parametric model is not very close to the OPD from the ground truth PSF, the resulting pixel polychromatic PSFs are very close.
- the pixel PSF decomposition between the parametric and the non-parametric part
- even if the model used to estimate the PSF is the same on used to generate the data, the case for the parametric Z=45 model, the Zernike coefficients are far from being close to the ones of the ground truth model.

All the plots are for one random test star.

## OPD

![OPDs](https://user-images.githubusercontent.com/16106854/109019624-704c2d80-76b9-11eb-8940-344b420d2d9e.png)

## Zernike coefficients 

![zernike coefficients](https://user-images.githubusercontent.com/16106854/109019771-9671cd80-76b9-11eb-9e1d-fd51e05b8652.png)

## OPD and PSFs

![psf_1](https://user-images.githubusercontent.com/16106854/109127185-6fb1a680-774e-11eb-9bf8-f0450cee4013.png)

![psf_2](https://user-images.githubusercontent.com/16106854/109127203-73452d80-774e-11eb-9db2-5ba918964a63.png)

![psf_3](https://user-images.githubusercontent.com/16106854/109127212-750ef100-774e-11eb-97ba-7a8e34714f51.png)
